File: Flutter_Realtime_Workspace/lib/app/domain/models/project_model.dart
class ProjectAttachment {
  final String url;
  final String? publicId;
  final String? resourceType;
  final String? format;
  final int? bytes;
  final String? filename;
  final String? originalFilename;
  final String? type;
  final int? width;
  final int? height;
  final double? duration;
  final DateTime? uploadedAt;
  final String? uploadedBy;

  const ProjectAttachment({
    required this.url,
    this.publicId,
    this.resourceType,
    this.format,
    this.bytes,
    this.filename,
    this.originalFilename,
    this.type,
    this.width,
    this.height,
    this.duration,
    this.uploadedAt,
    this.uploadedBy,
  });

  factory ProjectAttachment.fromJson(Map<String, dynamic> json) =>
      ProjectAttachment(
        url: json['url'] ?? '',
        publicId: json['public_id'] ?? json['publicId'],
        resourceType: json['resource_type'] ?? json['resourceType'],
        format: json['format'],
        bytes: json['bytes'],
        filename: json['filename'],
        originalFilename: json['original_filename'] ?? json['originalFilename'],
        type: json['type'],
        width: json['width'],
        height: json['height'],
        duration: (json['duration'] as num?)?.toDouble(),
        uploadedAt: json['uploadedAt'] != null
            ? DateTime.tryParse(json['uploadedAt'])
            : null,
        uploadedBy: json['uploadedBy'] is Map
            ? json['uploadedBy']['_id']
            : json['uploadedBy'],
      );

  Map<String, dynamic> toJson() => {
        'url': url,
        'public_id': publicId,
        'resource_type': resourceType,
        'format': format,
        'bytes': bytes,
        'filename': filename,
        'original_filename': originalFilename,
        'type': type,
        'width': width,
        'height': height,
        'duration': duration,
        'uploadedAt': uploadedAt?.toIso8601String(),
        'uploadedBy': uploadedBy,
      };
}

class ProjectTimelineEvent {
  final String title;
  final String? description;
  final DateTime date;
  final String? type;

  const ProjectTimelineEvent({
    required this.title,
    this.description,
    required this.date,
    this.type,
  });

  factory ProjectTimelineEvent.fromJson(Map<String, dynamic> json) =>
      ProjectTimelineEvent(
        title: json['title'] ?? '',
        description: json['description'],
        date: DateTime.tryParse(json['date'] ?? '') ?? DateTime.now(),
        type: json['type'],
      );

  Map<String, dynamic> toJson() => {
        'title': title,
        'description': description,
        'date': date.toIso8601String(),
        'type': type,
      };
}

class ProjectBudget {
  final double? allocated;
  final double? spent;
  final String currency;

  const ProjectBudget({
    this.allocated,
    this.spent,
    this.currency = 'USD',
  });

  factory ProjectBudget.fromJson(Map<String, dynamic> json) => ProjectBudget(
        allocated: (json['allocated'] as num?)?.toDouble(),
        spent: (json['spent'] as num?)?.toDouble(),
        currency: json['currency'] ?? 'USD',
      );

  Map<String, dynamic> toJson() => {
        'allocated': allocated,
        'spent': spent,
        'currency': currency,
      };
}

class ProjectTimeTracking {
  final double? estimated;
  final double? actual;
  final String unit; // hours

  const ProjectTimeTracking({
    this.estimated,
    this.actual,
    this.unit = 'hours',
  });

  factory ProjectTimeTracking.fromJson(Map<String, dynamic> json) =>
      ProjectTimeTracking(
        estimated: (json['estimated'] as num?)?.toDouble(),
        actual: (json['actual'] as num?)?.toDouble(),
        unit: json['unit'] ?? 'hours',
      );

  Map<String, dynamic> toJson() => {
        'estimated': estimated,
        'actual': actual,
        'unit': unit,
      };
}

class ProjectSettings {
  final bool isPublic;
  final bool allowComments;
  final bool notifications;
  final bool autoArchive;
  final int autoArchiveDays;

  const ProjectSettings({
    this.isPublic = false,
    this.allowComments = true,
    this.notifications = true,
    this.autoArchive = false,
    this.autoArchiveDays = 30,
  });

  factory ProjectSettings.fromJson(Map<String, dynamic> json) =>
      ProjectSettings(
        isPublic: json['isPublic'] ?? false,
        allowComments: json['allowComments'] ?? true,
        notifications: json['notifications'] ?? true,
        autoArchive: json['autoArchive'] ?? false,
        autoArchiveDays: json['autoArchiveDays'] ?? 30,
      );

  Map<String, dynamic> toJson() => {
        'isPublic': isPublic,
        'allowComments': allowComments,
        'notifications': notifications,
        'autoArchive': autoArchive,
        'autoArchiveDays': autoArchiveDays,
      };
}

class ProjectModel {
  final String id;
  final String tenantId; // ObjectId ref -> String (Organization)
  final String workspaceId;
  final String name;
  final String? key;
  final String? description;
  // template: Kanban | Scrum | Blank Project | Project Management | Task Tracking
  final String? template;
  // status: active | archived | on-hold | completed | planning | review | cancelled
  final String status;
  // priority: low | medium | high | critical
  final String priority;
  final String? color;
  final bool isActive;
  final bool archived;
  final bool completed;
  final bool recent;
  final bool starred;
  final String? teamId;
  final String createdBy;
  final List<String> collaborators;
  final List<String> members;
  final List<String> tags;
  final DateTime? startDate;
  final DateTime? endDate;
  final DateTime? lastViewed;
  final double progress; // 0 to 1
  final List<ProjectAttachment> attachments;
  final List<ProjectTimelineEvent> timeline;
  final ProjectBudget? budget;
  final ProjectTimeTracking? timeTracking;
  final ProjectSettings settings;
  final Map<String, dynamic>? customFields;
  final DateTime? deletedAt;
  final DateTime createdAt;
  final DateTime updatedAt;

  const ProjectModel({
    required this.id,
    required this.tenantId,
    required this.workspaceId,
    required this.name,
    this.key,
    this.description,
    this.template,
    this.status = 'active',
    this.priority = 'medium',
    this.color,
    this.isActive = true,
    this.archived = false,
    this.completed = false,
    this.recent = false,
    this.starred = false,
    this.teamId,
    required this.createdBy,
    this.collaborators = const [],
    this.members = const [],
    this.tags = const [],
    this.startDate,
    this.endDate,
    this.lastViewed,
    this.progress = 0,
    this.attachments = const [],
    this.timeline = const [],
    this.budget,
    this.timeTracking,
    this.settings = const ProjectSettings(),
    this.customFields,
    this.deletedAt,
    required this.createdAt,
    required this.updatedAt,
  });

  factory ProjectModel.fromJson(Map<String, dynamic> json) => ProjectModel(
        id: json['_id'] ?? json['id'] ?? '',
        tenantId: json['tenantId'] is Map
            ? json['tenantId']['_id'] ?? ''
            : json['tenantId'] ?? '',
        workspaceId: json['workspaceId'] is Map
            ? json['workspaceId']['_id'] ?? ''
            : json['workspaceId'] ?? '',
        name: json['name'] ?? '',
        key: json['key'],
        description: json['description'],
        template: json['template'],
        status: json['status'] ?? 'active',
        priority: json['priority'] ?? 'medium',
        color: json['color'],
        isActive: json['isActive'] ?? true,
        archived: json['archived'] ?? false,
        completed: json['completed'] ?? false,
        recent: json['recent'] ?? false,
        starred: json['starred'] ?? false,
        teamId: json['teamId'] is Map
            ? json['teamId']['_id']
            : json['teamId'],
        createdBy: json['createdBy'] is Map
            ? json['createdBy']['_id'] ?? ''
            : json['createdBy'] ?? '',
        collaborators: (json['collaborators'] as List? ?? [])
            .map((e) => e is Map ? e['_id'] as String? ?? '' : e as String)
            .toList(),
        members: (json['members'] as List? ?? [])
            .map((e) => e is Map ? e['_id'] as String? ?? '' : e as String)
            .toList(),
        tags: List<String>.from(json['tags'] ?? []),
        startDate: json['startDate'] != null
            ? DateTime.tryParse(json['startDate'])
            : null,
        endDate: json['endDate'] != null
            ? DateTime.tryParse(json['endDate'])
            : null,
        lastViewed: json['lastViewed'] != null
            ? DateTime.tryParse(json['lastViewed'])
            : null,
        progress: (json['progress'] as num?)?.toDouble() ?? 0,
        attachments: (json['attachments'] as List? ?? [])
            .map((a) =>
                ProjectAttachment.fromJson(a as Map<String, dynamic>))
            .toList(),
        timeline: (json['timeline'] as List? ?? [])
            .map((t) =>
                ProjectTimelineEvent.fromJson(t as Map<String, dynamic>))
            .toList(),
        budget: json['budget'] != null
            ? ProjectBudget.fromJson(json['budget'])
            : null,
        timeTracking: json['timeTracking'] != null
            ? ProjectTimeTracking.fromJson(json['timeTracking'])
            : null,
        settings: json['settings'] != null
            ? ProjectSettings.fromJson(json['settings'])
            : const ProjectSettings(),
        customFields: json['customFields'] as Map<String, dynamic>?,
        deletedAt: json['deletedAt'] != null
            ? DateTime.tryParse(json['deletedAt'])
            : null,
        createdAt:
            DateTime.tryParse(json['createdAt'] ?? '') ?? DateTime.now(),
        updatedAt:
            DateTime.tryParse(json['updatedAt'] ?? '') ?? DateTime.now(),
      );

  Map<String, dynamic> toJson() => {
        '_id': id,
        'tenantId': tenantId,
        'workspaceId': workspaceId,
        'name': name,
        'key': key,
        'description': description,
        'template': template,
        'status': status,
        'priority': priority,
        'color': color,
        'isActive': isActive,
        'archived': archived,
        'completed': completed,
        'recent': recent,
        'starred': starred,
        'teamId': teamId,
        'createdBy': createdBy,
        'collaborators': collaborators,
        'members': members,
        'tags': tags,
        'startDate': startDate?.toIso8601String(),
        'endDate': endDate?.toIso8601String(),
        'lastViewed': lastViewed?.toIso8601String(),
        'progress': progress,
        'attachments': attachments.map((a) => a.toJson()).toList(),
        'timeline': timeline.map((t) => t.toJson()).toList(),
        'budget': budget?.toJson(),
        'timeTracking': timeTracking?.toJson(),
        'settings': settings.toJson(),
        'customFields': customFields,
        'deletedAt': deletedAt?.toIso8601String(),
        'createdAt': createdAt.toIso8601String(),
        'updatedAt': updatedAt.toIso8601String(),
      };
}

File: Backend_Realtime_Workspace/modules/projects/models/project.model.js
// ============================================================================
// TeamSpot — Project Model
// Project management with timeline, attachments, budget, and time tracking
// ============================================================================

import mongoose from "mongoose";

const timelineEventSchema = new mongoose.Schema(
  {
    title: { type: String, required: true },
    description: { type: String },
    date: { type: Date, default: Date.now },
    type: { type: String, default: "custom" },
  },
  { _id: false }
);

const attachmentSchema = new mongoose.Schema(
  {
    url: { type: String, required: true },
    public_id: { type: String, required: true },
    resource_type: { type: String, required: true },
    format: { type: String },
    bytes: { type: Number },
    filename: { type: String, required: true },
    original_filename: { type: String },
    type: { type: String },
    width: { type: Number },
    height: { type: Number },
    duration: { type: Number },
    uploadedAt: { type: Date, default: Date.now },
    uploadedBy: { type: mongoose.Schema.Types.ObjectId, ref: "User" },
  },
  { _id: true }
);

const projectSchema = new mongoose.Schema(
  {
    // ── Multi-tenancy ─────────────────────────────────────────────────────
    tenantId: { type: mongoose.Schema.Types.ObjectId, ref: "Organization", required: true, index: true },
    workspaceId: { type: mongoose.Schema.Types.ObjectId, ref: "Workspace", index: true },

    // ── Core ──────────────────────────────────────────────────────────────
    name: { type: String, required: true, trim: true },
    key: { type: String, unique: true, sparse: true },
    description: { type: String, trim: true },
    template: { type: String, enum: ["Kanban", "Scrum", "Blank Project", "Project Management", "Task Tracking"], default: "Kanban" },
    status: { type: String, enum: ["active", "archived", "on-hold", "completed", "planning", "review", "cancelled"], default: "active" },
    priority: { type: String, enum: ["low", "medium", "high", "critical"], default: "medium" },
    color: { type: String, default: "#1E40AF" },

    // ── Flags ─────────────────────────────────────────────────────────────
    isActive: { type: Boolean, default: true },
    archived: { type: Boolean, default: false },
    completed: { type: Boolean, default: false },
    recent: { type: Boolean, default: false },
    starred: { type: Boolean, default: false },

    // ── Ownership ─────────────────────────────────────────────────────────
    teamId: { type: mongoose.Schema.Types.ObjectId, ref: "Team" },
    createdBy: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
    collaborators: [{ type: mongoose.Schema.Types.ObjectId, ref: "User" }],
    members: [{ type: mongoose.Schema.Types.ObjectId, ref: "User" }],
    tags: [{ type: String, trim: true, lowercase: true }],

    // ── Dates ─────────────────────────────────────────────────────────────
    startDate: { type: Date },
    endDate: { type: Date },
    lastViewed: { type: Date, default: Date.now },
    progress: { type: Number, min: 0, max: 1, default: 0 },

    // ── Attachments & Timeline ────────────────────────────────────────────
    attachments: [attachmentSchema],
    timeline: [timelineEventSchema],

    // ── Budget & Time ─────────────────────────────────────────────────────
    budget: {
      allocated: { type: Number, default: 0 },
      spent: { type: Number, default: 0 },
      currency: { type: String, default: "USD" },
    },
    timeTracking: {
      estimated: { type: Number, default: 0 },
      actual: { type: Number, default: 0 },
      unit: { type: String, default: "hours" },
    },

    // ── Settings ──────────────────────────────────────────────────────────
    settings: {
      isPublic: { type: Boolean, default: false },
      allowComments: { type: Boolean, default: true },
      notifications: { type: Boolean, default: true },
      autoArchive: { type: Boolean, default: false },
      autoArchiveDays: { type: Number, default: 90 },
    },
    customFields: { type: mongoose.Schema.Types.Mixed },

    // ── Soft delete ───────────────────────────────────────────────────────
    deletedAt: { type: Date },
  },
  { timestamps: true, toJSON: { virtuals: true }, toObject: { virtuals: true } }
);

// ── Indexes ───────────────────────────────────────────────────────────────
projectSchema.index({ tenantId: 1, teamId: 1, status: 1 });
projectSchema.index({ tenantId: 1, createdBy: 1 });
projectSchema.index({ tenantId: 1, starred: 1 });
projectSchema.index({ tenantId: 1, tags: 1 });
projectSchema.index({ lastViewed: -1 });
projectSchema.index({ name: "text", description: "text" });

// ── Virtuals ──────────────────────────────────────────────────────────────
projectSchema.virtual("progressPercentage").get(function () { return Math.round(this.progress * 100); });
projectSchema.virtual("attachmentCount").get(function () { return this.attachments?.length || 0; });
projectSchema.virtual("collaboratorCount").get(function () { return this.collaborators?.length || 0; });
projectSchema.virtual("isOverdue").get(function () { return this.endDate && new Date() > this.endDate && !this.completed; });

// ── Pre-save ──────────────────────────────────────────────────────────────
projectSchema.pre("save", function (next) {
  const oneWeekAgo = new Date();
  oneWeekAgo.setDate(oneWeekAgo.getDate() - 7);
  this.recent = this.lastViewed && this.lastViewed > oneWeekAgo;

  if (this.progress === 1 && !this.completed) { this.completed = true; this.status = "completed"; }
  else if (this.progress < 1 && this.completed) { this.completed = false; if (this.status === "completed") this.status = "active"; }
  next();
});

// ── Soft delete filter ────────────────────────────────────────────────────
projectSchema.pre(/^find/, function () {
  if (!this.getQuery().includeDeleted) this.where({ deletedAt: null });
});

// ── Methods ───────────────────────────────────────────────────────────────
projectSchema.methods.addTimelineEvent = function (title, description, type = "custom") {
  this.timeline.push({ title, description, date: new Date(), type });
  return this.save();
};

// ── Statics ───────────────────────────────────────────────────────────────
projectSchema.statics.getProjectStats = function (tenantId, teamId = null) {
  const match = { tenantId: new mongoose.Types.ObjectId(tenantId) };
  if (teamId) match.teamId = new mongoose.Types.ObjectId(teamId);
  return this.aggregate([
    { $match: match },
    { $group: { _id: null, total: { $sum: 1 }, active: { $sum: { $cond: [{ $eq: ["$status", "active"] }, 1, 0] } }, completed: { $sum: { $cond: [{ $eq: ["$status", "completed"] }, 1, 0] } }, averageProgress: { $avg: "$progress" } } },
  ]);
};

const Project = mongoose.model("Project", projectSchema);
export default Project;
